File: packages/oxlint-plugin-react-doctor/src/plugin/rules/design/no-ease-in-motion.ts
import { defineRule } from "../../utils/define-rule.js";
import { getAuthoritativeJsxAttribute } from "../../utils/get-authoritative-jsx-attribute.js";
import { getClassNameTokens } from "../../utils/get-class-name-tokens.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import { isProvenFramerMotionJsxElement } from "../../utils/is-proven-framer-motion-jsx-element.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { getInlineStyleExpression } from "./utils/get-inline-style-expression.js";
import { getStringFromClassNameAttr } from "./utils/get-string-from-class-name-attr.js";
import { getStylePropertyKey } from "./utils/get-style-property-key.js";
import { getStylePropertyStringValue } from "./utils/get-style-property-string-value.js";

const EASE_IN_TOKEN_PATTERN = /(?:^|[\s,])ease-in(?=$|[\s,])/i;
const TIMING_PROPERTY_NAMES = new Set([
  "transition",
  "transitionTimingFunction",
  "animation",
  "animationTimingFunction",
]);

const getMotionEaseProperty = (node: EsTreeNode): EsTreeNode | null => {
  if (!isNodeOfType(node, "Property") || getStylePropertyKey(node) !== "ease") return null;
  const easeValue = getStylePropertyStringValue(node);
  return easeValue === "easeIn" || easeValue === "ease-in" ? node : null;
};

export const noEaseInMotion = defineRule({
  id: "no-ease-in-motion",
  title: "UI motion starts with ease-in",
  severity: "warn",
  tags: ["design", "test-noise"],
  category: "Performance",
  recommendation:
    "Use ease-out for entrances and exits, or ease-in-out when an element remains visible while moving.",
  create: (context: RuleContext) => ({
    JSXAttribute(node: EsTreeNodeOfType<"JSXAttribute">) {
      const styleExpression = getInlineStyleExpression(node);
      if (styleExpression) {
        for (const property of styleExpression.properties ?? []) {
          const propertyName = getStylePropertyKey(property);
          const propertyValue = getStylePropertyStringValue(property);
          if (
            propertyName &&
            TIMING_PROPERTY_NAMES.has(propertyName) &&
            propertyValue &&
            EASE_IN_TOKEN_PATTERN.test(propertyValue)
          ) {
            context.report({
              node: property,
              message:
                "Ease-in delays the visible response and makes this interaction feel sluggish. Use ease-out or a responsive custom curve.",
            });
          }
        }
      }

      if (
        !isNodeOfType(node.name, "JSXIdentifier") ||
        node.name.name !== "transition" ||
        !isNodeOfType(node.parent, "JSXOpeningElement") ||
        !Object.is(getAuthoritativeJsxAttribute(node.parent.attributes, "transition"), node) ||
        !isProvenFramerMotionJsxElement(node.parent, context.scopes) ||
        !node.value ||
        !isNodeOfType(node.value, "JSXExpressionContainer") ||
        !isNodeOfType(node.value.expression, "ObjectExpression")
      ) {
        return;
      }
      const easeProperty = node.value.expression.properties?.find(getMotionEaseProperty);
      if (easeProperty) {
        context.report({
          node: easeProperty,
          message:
            "Ease-in makes the first part of this UI motion feel unresponsive. Prefer ease-out for state changes users trigger.",
        });
      }
    },
    JSXOpeningElement(node: EsTreeNodeOfType<"JSXOpeningElement">) {
      const classNameValue = getStringFromClassNameAttr(node);
      if (!classNameValue) return;
      if (!getClassNameTokens(classNameValue).includes("ease-in")) return;
      context.report({
        node,
        message:
          "This ease-in utility back-loads the visible response. Use ease-out or a purpose-built timing curve for UI motion.",
      });
    },
  }),
});

File: native/oxlint/rules/no-ease-in-motion.rs
use lazy_regex::{Lazy, Regex, lazy_regex};
use oxc_ast::AstKind;
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_span::GetSpan;

use crate::{
    AstNode,
    context::{ContextHost, LintContext},
    rule::Rule,
};

const INLINE_STYLE_MESSAGE: &str = "Ease-in delays the visible response and makes this interaction feel sluggish. Use ease-out or a responsive custom curve.";
const MOTION_MESSAGE: &str = "Ease-in makes the first part of this UI motion feel unresponsive. Prefer ease-out for state changes users trigger.";
const TAILWIND_MESSAGE: &str = "This ease-in utility back-loads the visible response. Use ease-out or a purpose-built timing curve for UI motion.";
const TIMING_PROPERTY_NAMES: [&str; 4] = [
    "transition",
    "transitionTimingFunction",
    "animation",
    "animationTimingFunction",
];
static EASE_IN_TOKEN_PATTERN: Lazy<Regex> = lazy_regex!(
    r"(?i)(?:^|[,\u{0009}\u{000A}\u{000B}\u{000C}\u{000D}\u{0020}\u{00A0}\u{1680}\u{2000}-\u{200A}\u{2028}\u{2029}\u{202F}\u{205F}\u{3000}\u{FEFF}])ease-in(?:$|[,\u{0009}\u{000A}\u{000B}\u{000C}\u{000D}\u{0020}\u{00A0}\u{1680}\u{2000}-\u{200A}\u{2028}\u{2029}\u{202F}\u{205F}\u{3000}\u{FEFF}])"
);

#[derive(Debug, Default, Clone)]
pub struct NoEaseInMotion;

declare_oxc_lint!(
    /// Disallow ease-in timing for interface motion.
    NoEaseInMotion,
    react_doctor_native,
    perf,
    version = "0.1.0",
    short_description = "Disallow ease-in timing for interface motion.",
);

impl Rule for NoEaseInMotion {
    fn should_run(&self, ctx: &ContextHost) -> bool {
        !is_non_production_file(ctx)
    }

    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        match node.kind() {
            AstKind::JSXAttribute(attribute) => {
                let Some(style) = get_inline_style_object_expression(attribute) else {
                    return;
                };
                for property_name in TIMING_PROPERTY_NAMES {
                    let Some((property, value)) =
                        get_effective_static_style_property_string_value(style, property_name)
                    else {
                        continue;
                    };
                    if EASE_IN_TOKEN_PATTERN.is_match(value) {
                        ctx.diagnostic(
                            OxcDiagnostic::warn(INLINE_STYLE_MESSAGE)
                                .with_label(property.span()),
                        );
                    }
                }
            }
            AstKind::JSXOpeningElement(opening_element) => {
                for transition_object in
                    get_static_motion_transition_objects(opening_element, ctx)
                {
                    let Some((ease_property, ease_value)) =
                        get_effective_static_style_property_string_value(transition_object, "ease")
                    else {
                        continue;
                    };
                    if matches!(ease_value, "easeIn" | "ease-in") {
                        ctx.diagnostic(
                            OxcDiagnostic::warn(MOTION_MESSAGE)
                                .with_label(ease_property.span()),
                        );
                    }
                }
                let Some(class_name) = get_static_class_name(opening_element) else {
                    return;
                };
                if tailwind_class_name_tokens(class_name)
                    .iter()
                    .any(|token| token.utility == "ease-in")
                {
                    ctx.diagnostic(
                        OxcDiagnostic::warn(TAILWIND_MESSAGE)
                            .with_label(opening_element.span),
                    );
                }
            }
            _ => {}
        }
    }
}
